java/arcs/core/util/platform/js/PlatformBigInt.kt

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.util
+import kotlin.js.JsName
+import kotlin.js.js
+
+@JsName("BigInt")
+external class JsBigInt(value: dynamic) {
+}
+
+@JsName("Number")
+external class JsNumber(value: dynamic) {
+}
+
+fun JsBigInt.toNumber() = JsNumber(this) as Double
+fun JsBigInt.toInt() = this.toNumber().toInt()
+
+fun jsbi(v: dynamic) = JsBigInt(v)
+
+class PlatformBigInt(val jsBigInt: JsBigInt) {
+    constructor(bigInt: String) : this(JsBigInt(bigInt))
+
+    fun toByte() = jsBigInt.toInt().toByte()
+
+    fun toChar() = jsBigInt.toInt().toChar()
+
+    fun toDouble() = jsBigInt.toNumber()
+
+    fun toFloat() = jsBigInt.toNumber().toFloat()
+
+    fun toInt() = jsBigInt.toInt()
+
+    fun toLong(): Long {
```

Comment:
It's not obvious to me what this does. A comment would be good. Ideally tests should be added as well, both to test what you have as well as to provide a place for easily adding regression tests if we find bugs in the future.

I understand that testing the JS codepath without setting up Kotlin -> JS compilation and compliant unit testing framework (I'm not sure if JUnit can run in JS compiled from Kotlin, probably depends on Java, doesn't it?) may be more pain than it's worth it. But then, arguably if we don't want to pay the price of testing "Kotlin -> JS" bindings, why are we introducing them?

Reply:
If you mean the code in toLong(), it's coercing a BigInteger (from JS) into a Long. Since JS doesn't have longs, we first bit-wise and it with a bigint representing 0xFFFFFFF (should be 2^32 -1 not 2^64-1) to get the lower 32 bits. Then we shift right the bigint by dividing by the bigint 2^32 (shouldn't be 2^64) and truncating to 32-bits. Finally, we OR the upper 32-bits and lower 32-bits to get a 64-bit Long.